File: go/fixture/go-mini/internal/services/alert_service.go
package services

import (
	"encoding/json"
	"fmt"
	"time"

	"example.com/go-mini/internal/models"
)

// Formatter formats alerts for the audit log.
type Formatter interface {
	Format(a models.Alert) string
	ContentType() string
}

type plainFormatter struct{}

func (plainFormatter) Format(a models.Alert) string {
	return a.Channel + " -> " + a.Recipient + ": " + a.Summary
}

func (plainFormatter) ContentType() string {
	return "text/plain"
}

type jsonFormatter struct{}

func (jsonFormatter) Format(a models.Alert) string {
	b, _ := json.Marshal(a)
	return string(b)
}

func (jsonFormatter) ContentType() string {
	return "application/json" //nolint:goconst // TODO
}

func newFormatter(kind string) Formatter { //nolint:ireturn // TODO
	if kind == "json" {
		return jsonFormatter{}
	}
	return plainFormatter{}
}

// AlertService is a service for alerts.
type AlertService struct {
	notifier  *Notifier
	audit     *AuditLog
	formatter Formatter
	raised    int
	recieved  int //nolint:misspell // TODO
	lastAt    time.Time
}

// NewAlertService creates a new AlertService.
func NewAlertService(notifier *Notifier, audit *AuditLog, format string) *AlertService {
	return &AlertService{
		notifier:  notifier,
		audit:     audit,
		formatter: newFormatter(format),
	}
}

// Raise raises the alert that matches the device's status.
func (s *AlertService) Raise(d models.Device) error {
	s.recieved++ //nolint:misspell // TODO
	s.lastAt = time.Now()
	var a models.Alert
	//nolint:goconst // TODO
	switch d.Status {
	case "READY":
		return nil
	case "DEGRADED":
		a = models.Alert{Channel: "slack", Recipient: "#fleet", Summary: "device " + d.ID + " is degraded"}
	case "DOWN":
		a = models.Alert{Channel: "pagerduty", Recipient: "fleet-oncall", Summary: "device " + d.ID + " is down"}
	default:
		return fmt.Errorf("no alert for status %q", d.Status)
	}
	s.raised++
	_ = s.audit.Write(s.formatter.ContentType() + " " + s.formatter.Format(a))
	return s.notifier.Send(a.Channel, a.Summary)
}

// Raised returns how many alerts have been raised.
func (s *AlertService) Raised() int {
	return s.raised
}

File: py/fixture/py-mini/internal/services/alert_service.py
"""Alerts raised from device status."""

import dataclasses
import json
from datetime import UTC, datetime
from typing import Protocol

from internal.models.alert import Alert
from internal.models.device import Device
from internal.services.device_service import AuditLog
from internal.services.notify import Notifier


class Formatter(Protocol):
    """Formats alerts for the audit log."""

    def format(self, a: Alert) -> str:
        """Render the alert."""
        ...

    def content_type(self) -> str:
        """Name the rendering's media type."""
        ...


class _PlainFormatter:
    def format(self, a: Alert) -> str:
        return a.channel + " -> " + a.recipient + ": " + a.summary

    def content_type(self) -> str:
        return "text/plain"


class _JsonFormatter:
    def format(self, a: Alert) -> str:
        return json.dumps(dataclasses.asdict(a))

    def content_type(self) -> str:
        return "application/json"


def _new_formatter(kind: str) -> Formatter:
    if kind == "json":  # noqa: PLR2004  # TODO
        return _JsonFormatter()
    return _PlainFormatter()


class AlertService:
    """AlertService is a service for alerts."""

    def __init__(self, notifier: Notifier, audit: AuditLog, fmt: str) -> None:
        """Create a new AlertService."""
        self.notifier = notifier
        self.audit = audit
        self.formatter = _new_formatter(fmt)
        self.raised = 0
        self.recieved = 0
        self.last_at: datetime | None = None

    def raise_alert(self, d: Device) -> None:
        """Raise the alert that matches the device's status."""
        self.recieved += 1
        self.last_at = datetime.now(UTC)
        match d.status:
            case "READY":
                return
            case "DEGRADED":
                a = Alert(
                    channel="slack", recipient="#fleet", summary="device " + d.id + " is degraded"
                )
            case "DOWN":
                a = Alert(
                    channel="pagerduty",
                    recipient="fleet-oncall",
                    summary="device " + d.id + " is down",
                )
            case _:
                raise ValueError(f'no alert for status "{d.status}"')
        self.raised += 1
        self.audit.write(self.formatter.content_type() + " " + self.formatter.format(a))
        self.notifier.send(a.channel, a.summary)

    def raised_count(self) -> int:
        """Return how many alerts have been raised."""
        return self.raised
